pkg/pillar/canbus/canbus.go

```diff
@@ -0,0 +1,881 @@
+
+	link.TDCV = tdcv
+	link.TDCO = tdco
+	link.TDCF = tdcf
+	return nil
+}
+
+// SetupCAN setup CAN interface from a map of properties described as
+// strings, e.g.:
+//
+//	properties := map[string]string{
+//	 "bitrate": "125000",
+//	 "sample-point": "0.875",
+//	 "tq": "29",
+//	 "prop_seg": "118",
+//	 "phase_seg1": "119",
+//	 "phase_seg2": "34",
+//	 "sjw": "1",
+//	}
+func SetupCAN(link *CAN, properties map[string]string) error {
+	if link == nil {
+		return errors.New("No link provided")
+	}
+	// Try to fetch current properties from the interface
+	var canProps CAN
+	iface, err := GetCANLink(link.Attrs().Name)
+	if err == nil {
+		canProps = *iface
+	} else {
+		canProps = CAN{}
+	}
+	changeBitTiming := false
+	changeDataBitTiming := false
+	changeCtrlMode := false
+	changeTermination := false
+	changeRestart := false
+	changeRestartMs := false
+	changeTDC := false
+	var cvErr [29]error
+
+	for propStr, valueStr := range properties {
+		prop := strings.ToLower(propStr)
+		value := strings.ToLower(valueStr)
+		switch prop {
+		case "bitrate":
+			canProps.BitTiming.Bitrate, cvErr[0] = convUint32(value)
+			changeBitTiming = true
+		case "sample-point":
+			canProps.BitTiming.Sample_point, cvErr[1] = convSamplePoint(value)
+			changeBitTiming = true
+		case "tq":
+			canProps.BitTiming.Tq, cvErr[2] = convUint32(value)
+			changeBitTiming = true
+		case "prop_seg":
+			canProps.BitTiming.Prop_seg, cvErr[3] = convUint32(value)
+			changeBitTiming = true
+		case "phase_seg1":
+			canProps.BitTiming.Phase_seg1, cvErr[4] = convUint32(value)
+			changeBitTiming = true
+		case "phase_seg2":
+			canProps.BitTiming.Phase_seg2, cvErr[5] = convUint32(value)
+			changeBitTiming = true
+		case "sjw":
+			canProps.BitTiming.Sjw, cvErr[6] = convUint32(value)
+			changeBitTiming = true
+		case "dbitrate":
+			canProps.DataBitTiming.Bitrate, cvErr[7] = convUint32(value)
+			changeDataBitTiming = true
+		case "dsample-point":
+			canProps.DataBitTiming.Sample_point, cvErr[8] = convSamplePoint(value)
+			changeDataBitTiming = true
+		case "dtq":
+			canProps.DataBitTiming.Tq, cvErr[9] = convUint32(value)
+			changeDataBitTiming = true
+		case "dprop_seg":
+			canProps.DataBitTiming.Prop_seg, cvErr[10] = convUint32(value)
+			changeDataBitTiming = true
+		case "dphase_seg1":
+			canProps.DataBitTiming.Phase_seg1, cvErr[11] = convUint32(value)
+			changeDataBitTiming = true
+		case "dphase_seg2":
+			canProps.DataBitTiming.Phase_seg2, cvErr[12] = convUint32(value)
+			changeDataBitTiming = true
+		case "dsjw":
+			canProps.DataBitTiming.Sjw, cvErr[13] = convUint32(value)
+			changeDataBitTiming = true
+		case "loopback":
+			cvErr[14] = setCtrlModeFlag(prop, value, &canProps.CtrlMode)
+			changeCtrlMode = true
+		case "listen-only":
+			cvErr[15] = setCtrlModeFlag(prop, value, &canProps.CtrlMode)
+			changeCtrlMode = true
+		case "triple-sampling":
+			cvErr[16] = setCtrlModeFlag(prop, value, &canProps.CtrlMode)
+			changeCtrlMode = true
+		case "one-shot":
```

Comment:
you parse `propStr` here and then you parse it in `setCtrlModeFlag` again - would be good having a single responsibility only

Reply:
I agree, the idea is that in SetupCAN all the parameters (passed as strings) are parsed. However, I've broke the logic to set Ctrl. mode to reduce the complexity of SetupCAN. Ctrl. mode represents just a small set of these parameters. My initial plan was to use the flag value itself to determine directly which flag we want to change, so it would be more easy to execute the switch. However, there is no a single flag to identify the CTRLMODE_TDC, so I decided to also use strings in order to no create more constants and increase the complexity of the code. These are suppose to be executed only when configuring the CAN interface, so I believe we can live with a "double" parsing of a small set of strings...

Reply:
we had a discussion yesterday @rene wanted to look at it again